app/helpers/federation_helper.rb

```diff
@@ -123,5 +123,63 @@ def federation_external_class?(class_object)
     !class_object.links['self'].include?($REST_URL)
   end
 
+  def canonical_ontology(ontologies)
+    if ontologies.size.eql?(1)
+      ontologies.first
+    else
+      internal_ontology = ontologies.select { |x| helpers.internal_ontology?(x[:id]) }.first
+      if internal_ontology
+        internal_ontology
+      else
+        external_canonical_ontology_portal(ontologies)
+      end
+    end
+  end
+
+  def external_canonical_ontology_portal(ontologies)
+    portal_counts = Hash.new(0)
+    # Count occurrences of each portal in the pull_location URL
+    ontologies.each do |ontology|
+      federated_portals.keys.each do |portal|
+        if ontology[:pullLocation]&.include?(portal.to_s)
+          portal_counts[portal] += 1
+        end
+      end
+    end
+    # Determine the portal with the most occurrences
+    portal = portal_counts.max_by { |_, count| count }&.first
+
+    ontologies.select{|o| o[:id].include?(portal.to_s)}.first
+  end
+
+  def apply_canonical_portal(search_results, all_submissions)
+    search_results.each do |result|
+      next if result[:root][:portal_name].nil? || result[:root][:other_portals].blank?
+
+      candidates = [result[:root][:link]] + result[:root][:other_portals].map { |p| p[:link].split('?').first }
+
+      portal_counts = Hash.new(0)
```

Comment:
from lines `161` to `170`  is the same as in line `139` in the function `external_canonical_ontology_portal`

Reply:
Yes We can extract that code in a function`count_portal_occurrences` but this abstraction will make the code less readable. 

We can do something like this:
```
  def count_portal_occurrences(items, key_extractor)
    portal_counts = Hash.new(0)
    items.each do |item|
      federated_portals.keys.each do |portal|
        portal_counts[portal] += 1 if key_extractor.call(item)&.include?(portal.to_s)
      end
    end
    portal_counts.max_by { |_, count| count }&.first
  end
```

and `key_extractor` is a lambda function.

But I see no need to complicate this code more.

Reply:
you can also have a function that has as input ontologies IDs, and for the search case, you map the results to ontologies IDs. I did something similar here https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/778/commits/ac1f608256472b59ed7b1c24cf3c2bd51657f70e and https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/778/commits/a653fd1342cb941a8b0250e85e7c0a9a7f90819f